### PR TITLE
In Windows host, use WriteConsoleW for stdout and stderr, and locale enabled print for file output

### LINKED ISSUE #17451: CLI's --list-runtimes doesn't work well with UTF-8 characters
Hello,

My product relies on the .NET CLI for some operations, and specifically parses the output of the "dotnet --list-runtimes" command.
This doesn't work well however whenever a runtime is present on a path containing UTF-8 characters, as the CLI seem to convert these characters to something else:
image

.NET CLI version: 5.0.202

### LINKED ISSUE #102174: GB1030 dotnet info
1. copy a group of GB18030 characters(level 3 long sample data)
𫚭﹫𪛒𫍲𫟼𬘭𮯠⾢)

2. Use the supported string to create folder.  龯蝌灋齅ㄥ䶱

3. unzip dotnet-sdk-9.0.100-preview.4.24218.26-win-x64.zip to above folder
4. access to the folder and run `dotnet --info`

Expected Result:

the contents can be displayed well.

Actual Result:

GB18030 characters displays incorrectly.
image

### PR DESCRIPTION
**Pull request by jtschuster**
Fixes https://github.com/dotnet/runtime/issues/102174

Per https://github.com/dotnet/runtime/issues/102174#issuecomment-2109289546, this uses WriteConsoleW for writing to stdout and stderr, and the locale enabled _vfwprintf_l for writing to a file.

### CODE REVIEW

File: src/native/corehost/hostmisc/pal.h [hunk 1]
```diff
@@ -159,13 +159,48 @@ namespace pal
...
     inline FILE* file_open(const string_t& path, const char_t* mode) { return ::_wfsopen(path.c_str(), mode, _SH_DENYNO); }
 
-    inline void file_vprintf(FILE* f, const char_t* format, va_list vl) { ::vfwprintf(f, format, vl); ::fputwc(_X('\n'), f); }
-    inline void err_fputs(const char_t* message) { ::fputws(message, stderr); ::fputwc(_X('\n'), stderr); }
-    inline void out_vprintf(const char_t* format, va_list vl) { ::vfwprintf(stdout, format, vl); ::fputwc(_X('\n'), stdout); }
-
     inline int str_vprintf(char_t* buffer, size_t count, const char_t* format, va_list vl) { return ::_vsnwprintf_s(buffer, count, _TRUNCATE, format, vl); }
     inline int strlen_vprintf(const char_t* format, va_list vl) { return ::_vscwprintf(format, vl); }
 
+    inline void file_vprintf(FILE* f, const char_t* format, va_list vl)
```
[the remark this answers is not in the document]
Comment (jtschuster) [reply src/native/corehost/hostmisc/pal.h:165]: Is there a way to move them and keep them `inline`? I couldn't find a way to, so I've moved them and made them non-inline for now.

File: src/native/corehost/hostmisc/pal.h [hunk 2]
```diff
@@ -159,13 +159,48 @@ namespace pal
...
+
+    inline void err_fputs(const char_t* message) {
+        HANDLE e = GetStdHandle(STD_ERROR_HANDLE);
+        WriteConsoleW(e, message, (int)strlen(message), NULL, NULL);
+        WriteConsoleW(e, _X("\n"), 1, NULL, NULL);
+    }
+
+    inline void out_vprintf(const char_t* format, va_list vl) {
+        // Get the length of the formatted string + newline + null terminator
+        int len = 2 + strlen_vprintf(format, vl);
```
[the remark this answers is not in the document]
Comment (jtschuster) [reply src/native/corehost/hostmisc/pal.h:181]: Ah, I think so, that's probably what's causing the test failures.

**Comment by jtschuster** [comment]
Looks like the tests are failing because [WriteConsole fails](https://learn.microsoft.com/en-us/windows/console/writeconsole#:~:text=WriteConsole%20fails%20if%20it%20is%20used%20with%20a%20standard%20handle%20that%20is%20redirected%20to%20a%20file.) when stdout or stderr is redirected, so we need to write to a file in that case. Working on the fix.

**Comment by jtschuster** [comment]
There's a number of other string related functions in the pal that have locale variants, is there any reason not to update those to use UTF-8 everywhere in a follow-up?

**Review by elinor-fung — COMMENTED** [review]
Is there a simple test we could add for this?
Maybe something similar to the below, but using `DotNetBuilder` to set up the install in a directory with characters that don't work properly without the change in this PR: 
https://github.com/dotnet/runtime/blob/8b1d06a5170628834b27c06b751ac0e54f65c7e3/src/installer/tests/HostActivation.Tests/DotnetArgValidation.cs#L87

**Review by elinor-fung — APPROVED** [review]
Thanks!

**Pull request merged**